lib/matroid/connection.rb

```diff
@@ -0,0 +1,125 @@
+require 'base64'
+require 'json'
+require 'restclient'
+require 'date'
+
+BASE_API_URI       = 'https://www.matroid.com/api/0.1'
+DEFAULT_GRANT_TYPE = 'client_credentials'
+TOKEN_RESOURCE     = '/oauth/token'
+VERBS              = %w(get post)
+
+# error messages
+NO_ENV_VARIABLES_ERR     = 'AuthenticationError: "MATROID_CLIENT_ID" and "MATROID_CLIENT_SECRET" not found in environment'
+BAD_ENV_VARIABLES_ERR    = 'AuthenticationError: problem using environment variables "MATROID_CLIENT_ID" and "MATROID_CLIENT_SECRET"'
+BAD_CLIENT_VARIABLES_ERR = 'AuthenticationError: problem using client variables provided'
+
+module Matroid
+
+  # @attr_reader [Token] The current stored token object
+  class << self
+    attr_reader :token
+
+    VERBS.each do |verb|
+      define_method verb do |endpoint, *params|
+        send_request(verb, endpoint, *params)
+      end
+    end
+
+    private
+
+    def get_token(client_id, client_secret)
+      url = BASE_API_URI + TOKEN_RESOURCE
+      params = auth_params(client_id, client_secret)
+      begin
+        response = RestClient.post(url, params)
+      rescue RestClient::ExceptionWithResponse => e
+        puts JSON.pretty_generate(JSON.parse(e.response))
+        @token = nil
+        return false
+      end
+
+      @token = Token.new(JSON.parse(response))
+    end
+
+    def send_request(verb, path, *params)
+      url = path.start_with?('http') ? path : BASE_API_URI + path
+      url, query = *url.split('?')
```

Comment:
There could be [more than one question mark](http://stackoverflow.com/questions/2924160/is-it-valid-to-have-more-than-one-question-mark-in-a-url) in the url. [CGI](http://ruby-doc.org/stdlib-2.4.1/libdoc/cgi/rdoc/CGI.html#M000075) might be helpful here.